File: choreolib/src/main/java/com/choreo/lib/Choreo.java
// Copyright (c) Choreo contributors

package com.choreo.lib;

import static edu.wpi.first.util.ErrorMessages.requireNonNullParam;

import com.choreo.lib.trajectory.ChoreoTrajectory;
import com.google.gson.Gson;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Filesystem;
import edu.wpi.first.wpilibj2.command.Subsystem;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import java.util.function.Supplier;

/** Utilities to load and follow ChoreoTrajectories */
public final class Choreo {
  private static final Gson gson = new Gson();

  /** Default constructor. */
  private Choreo() {
    throw new UnsupportedOperationException("This is a utility class!");
  }

  private static Optional<ChoreoTrajectory> loadFile(File path) {
    try {
      var reader = new BufferedReader(new FileReader(path));
      ChoreoTrajectory traj = gson.fromJson(reader, ChoreoTrajectory.class);
      return Optional.ofNullable(traj);
    } catch (FileNotFoundException ex) {
      return Optional.empty();
    } catch (Exception ex) {
      DriverStation.reportError(ex.getMessage(), ex.getStackTrace());
    }
    return Optional.empty();
  }

  /**
   * Load a trajectory from the deploy directory. Choreolib expects .traj files to be placed in
   * src/main/deploy/choreo/[trajName].traj.
   *
   * @param trajName the path name in Choreo, which matches the file name in the deploy directory.
   * @return the loaded trajectory, or null if the trajectory could not be loaded.
   */
  public static Optional<ChoreoTrajectory> getTrajectory(String trajName) {
    requireNonNullParam(trajName, "trajName", "Choreo.getTrajectory");

    String fileExtension = ".traj";
    if (trajName.endsWith(fileExtension)) {
      trajName = trajName.substring(0, trajName.length() - fileExtension.length());
    }
    var traj_dir = new File(Filesystem.getDeployDirectory(), "choreo");
    var traj_file = new File(traj_dir, trajName + fileExtension);

    return loadFile(traj_file);
  }

  /**
   * Loads the split parts of the specified trajectory. Fails and returns null if any of the parts
   * could not be loaded.
   *
   * <p>This method determines the number of parts to load by counting the files that match the
   * pattern "trajName.X.traj", where X is a string of digits. Let this count be N. It then attempts
   * to load "trajName.1.traj" through "trajName.N.traj", consecutively counting up. If any of these
   * files cannot be loaded, the method returns null.
   *
   * @param trajName The path name in Choreo for this trajectory.
   * @return The List of segments, in order, or null.
   */
  public static Optional<List<ChoreoTrajectory>> getTrajectoryGroup(String trajName) {
    requireNonNullParam(trajName, "trajName", "Choreo.getTrajectoryGroup");

    // Count files matching the pattern for split parts.
    var traj_dir = new File(Filesystem.getDeployDirectory(), "choreo");
    File[] files =
        traj_dir.listFiles((file) -> file.getName().matches(trajName + "\\.\\d+\\.traj"));
    int segmentCount = files.length;
    // Try to load the segments.
    var trajs = new ArrayList<ChoreoTrajectory>();
    for (int i = 1; i <= segmentCount; ++i) {
      File traj = new File(traj_dir, String.format("%s.%d.traj", trajName, i));
      var trajectory = loadFile(traj);
      if (!trajectory.isPresent()) {
        DriverStation.reportError("ChoreoLib: Missing segments for path group " + trajName, false);
        return Optional.empty();
      }
      trajs.add(trajectory.get());
    }

    return Optional.of(trajs);
  }

  /**
   * Create a command to follow a Choreo path.
   *
   * @param driveSubsystem The drive subsystem to require for commands made from this factory.
   * @param poseSupplier A function that returns the current field-relative pose of the robot.
   * @param controller A ChoreoControlFunction to follow the current trajectory state. Use
   *     ChoreoCommands.choreoSwerveController(PIDController xController, PIDController yController,
   *     PIDController rotationController) to create one using PID controllers for each degree of
   *     freedom. You can also pass in a function with the signature (Pose2d currentPose,
   *     ChoreoTrajectoryState referenceState) -&gt; ChassisSpeeds to implement a custom follower
   *     (i.e. for logging).
   * @param outputChassisSpeeds A function that consumes the target robot-relative chassis speeds
   *     and commands them to the robot.
   * @param mirrorTrajectory If this returns true, the path will be mirrored to the opposite side,
   *     while keeping the same coordinate system origin. This will be called every loop during the
   *     command.
   * @return A command that follows a Choreo path.
   */
  public static ChoreoAutoFactory createAutoFactory(
      Subsystem driveSubsystem,
      Supplier<Pose2d> poseSupplier,
      ChoreoControlFunction controller,
      Consumer<ChassisSpeeds> outputChassisSpeeds,
      BooleanSupplier mirrorTrajectory) {
    return new ChoreoAutoFactory(
        requireNonNullParam(poseSupplier, "poseSupplier", "Choreo.createAutoFactory"),
        requireNonNullParam(controller, "controller", "Choreo.createAutoFactory"),
        requireNonNullParam(outputChassisSpeeds, "outputChassisSpeeds", "Choreo.createAutoFactory"),
        requireNonNullParam(mirrorTrajectory, "mirrorTrajectory", "Choreo.createAutoFactory"),
        Optional.empty(),
        requireNonNullParam(driveSubsystem, "driveSubsystem", "Choreo.createAutoFactory"));
  }

  /**
   * Create a command to follow a Choreo path.
   *
   * @param driveSubsystem The drive subsystem to require for commands made from this factory.
   * @param poseSupplier A function that returns the current field-relative pose of the robot.
   * @param controller A ChoreoControlFunction to follow the current trajectory state. Use
   *     ChoreoCommands.choreoSwerveController(PIDController xController, PIDController yController,
   *     PIDController rotationController) to create one using PID controllers for each degree of
   *     freedom. You can also pass in a function with the signature (Pose2d currentPose,
   *     ChoreoTrajectoryState referenceState) -&gt; ChassisSpeeds to implement a custom follower
   *     (i.e. for logging).
   * @param outputChassisSpeeds A function that consumes the target robot-relative chassis speeds
   *     and commands them to the robot.
   * @param mirrorTrajectory If this returns true, the path will be mirrored to the opposite side,
   *     while keeping the same coordinate system origin. This will be called every loop during the
   *     command.
   * @param trajLogger A function that consumes a trajectory whever one is started, should be used
   *     for logging.x
   * @return A command that follows a Choreo path.
   */
  public static ChoreoAutoFactory createAutoFactory(
      Subsystem driveSubsystem,
      Supplier<Pose2d> poseSupplier,
      ChoreoControlFunction controller,
      Consumer<ChassisSpeeds> outputChassisSpeeds,
      BooleanSupplier mirrorTrajectory,
      Consumer<ChoreoTrajectory> trajLogger) {
    return new ChoreoAutoFactory(
        requireNonNullParam(poseSupplier, "poseSupplier", "Choreo.createAutoFactory"),
        requireNonNullParam(controller, "controller", "Choreo.createAutoFactory"),
        requireNonNullParam(outputChassisSpeeds, "outputChassisSpeeds", "Choreo.createAutoFactory"),
        requireNonNullParam(mirrorTrajectory, "mirrorTrajectory", "Choreo.createAutoFactory"),
        Optional.of(requireNonNullParam(trajLogger, "trajLogger", "Choreo.createAutoFactory")),
        requireNonNullParam(driveSubsystem, "driveSubsystem", "Choreo.createAutoFactory"));
  }
}

File: choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp
// Copyright (c) Choreo contributors

#include "choreo/lib/Choreo.h"

#include <filesystem>
#include <memory>
#include <numbers>
#include <regex>
#include <string>
#include <vector>

#include <fmt/format.h>
#include <frc/DriverStation.h>
#include <frc/Filesystem.h>
#include <frc/Timer.h>
#include <frc2/command/FunctionalCommand.h>
#include <wpi/MemoryBuffer.h>
#include <wpi/json.h>

#include "choreo/lib/ChoreoSwerveCommand.h"

using namespace choreolib;

ChoreoTrajectory Choreo::GetTrajectory(std::string_view trajectoryName) {
  std::string trajectoryFileName =
      fmt::format("{}/choreo/{}.traj", frc::filesystem::GetDeployDirectory(),
                  trajectoryName);

  auto fileBuffer = wpi::MemoryBuffer::GetFile(trajectoryFileName);
  if (!fileBuffer) {
    throw std::runtime_error(
        fmt::format("Cannot open file: {}", trajectoryFileName));
  }

  wpi::json json = wpi::json::parse(fileBuffer.value()->GetCharBuffer());

  ChoreoTrajectory trajectory;
  choreolib::from_json(json, trajectory);
  return trajectory;
}

std::vector<ChoreoTrajectory> Choreo::GetTrajectoryGroup(
    std::string_view trajectoryName) {
  const std::filesystem::path trajectoryDir{
      fmt::format("{}/choreo", frc::filesystem::GetDeployDirectory())};
  int segmentCount = 0;
  for (const auto& dir_entry :
       std::filesystem::directory_iterator{trajectoryDir}) {
    if (dir_entry.is_regular_file() &&
        std::regex_match(
            dir_entry.path().filename().string(),
            std::regex(fmt::format("{}\\.\\d+\\.traj", trajectoryName)))) {
      ++segmentCount;
    }
  }
  std::vector<ChoreoTrajectory> group;
  group.reserve(segmentCount);
  for (int i = 1; i <= segmentCount; ++i) {
    try {
      group.push_back(
          Choreo::GetTrajectory(fmt::format("{}.{}", trajectoryName, i)));
    } catch (const std::exception&) {
      throw std::runtime_error(
          fmt::format("Cannot open file: {}.{}.traj", trajectoryName, i));
    }
  }
  return group;
}

frc2::CommandPtr Choreo::ChoreoSwerveCommandFactory(
    ChoreoTrajectory trajectory, std::function<frc::Pose2d()> poseSupplier,
    frc::PIDController xController, frc::PIDController yController,
    frc::PIDController rotationController,
    std::function<void(frc::ChassisSpeeds)> outputChassisSpeeds,
    std::function<bool(void)> mirrorTrajectory,
    frc2::Requirements requirements) {
  return ChoreoSwerveCommand(trajectory, poseSupplier,
                             ChoreoSwerveController(xController, yController,
                                                    rotationController),
                             outputChassisSpeeds, mirrorTrajectory,
                             requirements)
      .ToPtr();
}

frc2::CommandPtr Choreo::ChoreoSwerveCommandFactory(
    ChoreoTrajectory trajectory, std::function<frc::Pose2d()> poseSupplier,
    ChoreoControllerFunction controller,
    std::function<void(frc::ChassisSpeeds)> outputChassisSpeeds,
    std::function<bool(void)> mirrorTrajectory,
    frc2::Requirements requirements) {
  return ChoreoSwerveCommand(trajectory, poseSupplier, controller,
                             outputChassisSpeeds, mirrorTrajectory,
                             requirements)
      .ToPtr();
}

ChoreoControllerFunction Choreo::ChoreoSwerveController(
    frc::PIDController xController, frc::PIDController yController,
    frc::PIDController rotationController) {
  rotationController.EnableContinuousInput(-std::numbers::pi, std::numbers::pi);
  return [xController, yController, rotationController](
             frc::Pose2d pose, ChoreoTrajectoryState referenceState) mutable {
    units::meters_per_second_t xFF = referenceState.velocityX;
    units::meters_per_second_t yFF = referenceState.velocityY;
    units::radians_per_second_t rotationFF = referenceState.angularVelocity;

    units::meters_per_second_t xFeedback{
        xController.Calculate(pose.X().value(), referenceState.x.value())};
    units::meters_per_second_t yFeedback{
        yController.Calculate(pose.Y().value(), referenceState.y.value())};
    units::radians_per_second_t rotationFeedback{rotationController.Calculate(
        pose.Rotation().Radians().value(), referenceState.heading.value())};

    return frc::ChassisSpeeds::FromFieldRelativeSpeeds(
        xFF + xFeedback, yFF + yFeedback, rotationFF + rotationFeedback,
        pose.Rotation());
  };
}
